skip embedding_bag since its backward is not compatible to cudagraph

brad-mengchi opened the pull request:
Summary:
Work around that the bwd of embedding_bag doesn't work with cudagraph with the following error:

RuntimeError: unique_by_key: failed to synchronize: cudaErrorStreamCaptureUnsupported: operation not permitted when stream is capturing

and then cuda memory error

Differential Revision: D40625081



cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

desertfire commented:
What happens if we force `embedding_bag` to be a fallback in inductor and turns off cudagraph in inductor? Can it solve your problem here?

brad-mengchi committed:
skip embedding_bag since its backward is not compatible to cudagraph (#88874)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88874

Work around that the bwd of embedding_bag doesn't work with cudagraph with the following error:

RuntimeError: unique_by_key: failed to synchronize: cudaErrorStreamCaptureUnsupported: operation not permitted when stream is capturing

and then cuda memory error

Differential Revision: D40625081

fbshipit-source-id: 8950bf123cb52b90d86179fea7d35b9d9c8935b7

anijain2305 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -172,6 +173,11 @@ def call_function(
     ) -> "VariableTracker":
         from . import ConstantVariable, GradModeVariable, TensorVariable
 
+        if inductor_config.triton.cudagraphs and self.value.__name__ in [
```
This does not look like the right place to enter this code.

There are a couple of ways

1) At user level - use `torch._dynamo.disallow_in_graph` - https://github.com/pytorch/pytorch/blob/master/test/dynamo/test_misc.py#L1696-L1714. This would cause a graph break at dynamo-level w/o changing Dynamo code. This gives the control to user.

2) At inductor level - Check if the fx graph has an embedding_op, if it has skip cudagraphs. The relevant guarding is here - https://github.com/pytorch/pytorch/blob/master/torch/_inductor/compile_fx.py#L130-L133. This is not a good option if the graph is very large, and we have to skip cudagraphs for the whole graph just because of embedding_op.
